modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/compute/ClientComputeExecuteRequest.java

```diff
@@ -50,29 +56,82 @@ public class ClientComputeExecuteRequest {
     public static CompletableFuture<Void> process(
             ClientMessageUnpacker in,
             ClientMessagePacker out,
-            IgniteCompute compute,
+            IgniteComputeInternal compute,
             ClusterService cluster,
-            NotificationSender notificationSender) {
-        var nodeName = in.tryUnpackNil() ? null : in.unpackString();
-
-        var node = nodeName == null
-                ? cluster.topologyService().localMember()
-                : cluster.topologyService().getByConsistentId(nodeName);
-
-        if (node == null) {
-            throw new IgniteException("Specified node is not present in the cluster: " + nodeName);
-        }
+            NotificationSender notificationSender
+    ) {
+        Set<ClusterNode> candidates = unpackCandidateNodes(in, cluster);
 
         List<DeploymentUnit> deploymentUnits = unpackDeploymentUnits(in);
         String jobClassName = in.unpackString();
         JobExecutionOptions options = JobExecutionOptions.builder().priority(in.unpackInt()).maxRetries(in.unpackInt()).build();
         Object[] args = unpackArgs(in);
 
-        JobExecution<Object> execution = compute.executeAsync(Set.of(node), deploymentUnits, jobClassName, options, args);
+        ClusterNode localNode = cluster.topologyService().localMember();
+        ClusterNode targetNode = selectTargetNode(candidates, localNode);
+        candidates.remove(targetNode);
+
+        JobExecution<Object> execution = compute.executeAsyncWithFailover(targetNode, candidates,
+                deploymentUnits, jobClassName, options, args);
         sendResultAndStatus(execution, notificationSender);
         return execution.idAsync().thenAccept(out::packUuid);
     }
 
+    private static Set<ClusterNode> unpackCandidateNodes(ClientMessageUnpacker in, ClusterService cluster) {
+        Set<String> candidateIds = Arrays.stream(in.unpackObjectArrayFromBinaryTuple())
+                .map(String.class::cast)
+                .collect(Collectors.toSet());
+
+        Set<ClusterNode> candidates = new HashSet<>();
+        List<String> notFoundNodes = new ArrayList<>();
+        candidateIds.forEach(id -> {
+            ClusterNode node = cluster.topologyService().getByConsistentId(id);
+            if (node == null) {
+                notFoundNodes.add(id);
+            } else {
+                candidates.add(node);
+            }
+        });
+
+        if (!notFoundNodes.isEmpty()) {
+            throw new IgniteException("Specified nodes are not present in the cluster: " + notFoundNodes);
+        }
+        return candidates;
+    }
+
+    /**
+     * Selects a random node from the set of candidates, preferably not a local node.
```

Comment:
Why do we prefer a non-local node?

Reply:
Sorry, I still don't understand. All nodes have a similar chance of failing, how is the local one different?

Reply:
* It is faster to run the job on the current node (no need for a network hop)
* Client logic specifically tries to send the job to the target node directly and avoid a network hop, but this logic does the opposite and tries to add a network hop back - does not make sense
* Node failure is relatively rare
* Client has retry mechanism to deal with failures

Let's invert the logic and invoke locally when possible.

Reply:
@Pochatkin Sending the job to a different node does not affect the chance of coordinator failure. 

If the coordinator fails, the job will fail for the client anyway (due to connection loss). And the client will retry it.

Reply:
In other words:
* Let's say node A is _destined_ to fail very soon
* Client sends the job to the node A. 

Variant 1 (your logic): 
* Node A sends the job to node B
* Node A fails, client sees broken connection, and considers the job failed

Variant 2:
* Node A executes the job locally
* Node A fails, client sees broken connection, and considers the job failed

The outcome is identical.

Reply:
Agree with this, client retry is different and can even be disabled. 

Still, using a remote node instead of a local one for job execution does not make sense. It degrades performance and does not improve anything.


I think we should design client-side compute job failover. The design goals can't be achieved with "coordinator node" approach, because coordinator can fail, or the connection to coordinator can be briefly lost (and re-established). The client should be the coordinator.

Reply:
@Pochatkin please read my previous replies. The single point of failure is not eliminated by sending the job to another node.